### PR TITLE
merge #12752, #10785, #17896, #17957, #18021, #18112: serialization improvements 

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Overview

As part of https://github.com/dashpay/dash/pull/4025, component pull requests that are required by TorV3 logic are being broken down into branches wherever feasable. This PR deals with serialization improvements taken from https://github.com/bitcoin/bitcoin/pull/10785 (split into #17850, #17896, #17957, #18021, #18112 and others)

## Contents

* Merge #17850 ("Serialization improvements (minimal initial commits)" by sipa)
  *  ~~9250a087d2f450f37342010aea0a5d583eea508b has been skipped for the sake of brevity, not relevant to https://github.com/dashpay/dash/pull/4025~~ Resolved in 7396f35 
* Merge #17896 ("Serialization improvements (step 2)" by sipa)
  * ~~9b66083788581c264a097e26795561cb3eac455d has been skipped for reasons mentioned above~~ Resolved in bbe9c25 
* Merge of #17957 ("Serialization improvements step 3 (compression.h)" by sipa)
  * ~~4de934b9b5b4be1bac8fe205f4ee9a79e772dc34 has been skipped as it presumes the presence of https://github.com/bitcoin/bitcoin/commit/d2d7267e235a3778544f3d472763c61dbdc19b00~~ Resolved in 6e12999
* Merge #18021 ("Serialization improvements step 4 (undo.h)" by sipa)
  * ~~3c94b0039d2ca2a8c41fd6127ff5019a2afc304e is skipped~~ Resolved in 5bd678d 
* Merge #18112 ("Serialization improvements (minimal initial commits)" by sipa)
  * ~~353f376277ad9b87e03c9ccbc1028c4b6d12e8ea skpd~~ Resolved in 9036875

### CODE REVIEW

**Review by UdjinM6 — DISMISSED** [review]
Not sure if I like skipping commits here, would be nice to finalize these backports while we are at it... But LGTM otherwise, so

utACK

**Review by xdustinface — CHANGES_REQUESTED** [review]
One nit see below and i also feel like we shouldn't just randomly skip commits while backporting PRs only because they are "not relevant". There is imo no good reason to only partially backport PRs if its possible to completely merge them in.. skipping parts of them just makes reviewing the PR as a whole a bit weird/confusing here and there imo and this even two times, now and when backporting the remaining parts.

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/chain.h [hunk 1]
```diff
@@ -383,33 +381,25 @@ class CDiskBlockIndex : public CBlockIndex
...
-        READWRITE(hash);
+        if (!(s.GetType() & SER_GETHASH)) READWRITE(VARINT(_nVersion, VarIntMode::NONNEGATIVE_SIGNED));
+
+        READWRITE(VARINT(obj.nHeight, VarIntMode::NONNEGATIVE_SIGNED));
+        READWRITE(VARINT(obj.nStatus));
+        READWRITE(VARINT(obj.nTx));
+        if (obj.nStatus & (BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO)) READWRITE(VARINT(obj.nFile, VarIntMode::NONNEGATIVE_SIGNED));
+        if (obj.nStatus & BLOCK_HAVE_DATA) READWRITE(VARINT(obj.nDataPos));
+        if (obj.nStatus & BLOCK_HAVE_UNDO) READWRITE(VARINT(obj.nUndoPos));
+
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/chain.h:395]: Good catch. Resolved in cf52d88!

**Review by UdjinM6 — DISMISSED** [review]
slightly tested ACK

**Review by xdustinface — DISMISSED** [review]
utACK

**Comment by PastaPastaPasta** [comment]
please resolve conflicts

**Comment by PastaPastaPasta** [comment]
@UdjinM6 is this merge okay, or does it need to be a rebase?

**Comment by UdjinM6** [comment]
¯\\\_(ツ)_/¯

Personally, I'd prefer a rebase but technically this merge should be ok too I guess

**Comment by PastaPastaPasta** [comment]
Yeah, I prefer rebase I think. @kittywhiskers please see https://github.com/PastaPastaPasta/dash/commits/pr_4027 I rebased on develop and squashed your commits

**Comment by kittywhiskers** [comment]
Branch updated with @PastaPastaPasta's fork as-is

**Review by PastaPastaPasta — APPROVED** [review]
utACK, although should probably adjust title

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**